### ISSUE TITLE
change build tools to migrate Java translations to babel 

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/chipper#132]
PropertiesToJSON and XMLToJSON live in chipper/grunt/, and they are used to migrate Java translations to HTML5 sims. They currently migrate to the strings/ directory in the sim's repo.  They should be changed to migrate to the sim's directory in babel.

Assigning to @jonathanolson, since it may be easy to address this while doing https://github.com/phetsims/chipper/issues/131.

**Comment by jonathanolson** [comment]
Taking care of https://github.com/phetsims/chipper/issues/131 (just had to write proper encoding for converting locales), so I'll be using plinko-probability for the Flash conversion and states-of-mattter for the Java conversion.

**Comment by jonathanolson** [comment]
Done with https://github.com/phetsims/chipper/issues/131#issuecomment-104822458, will wait until I can commit before closing.

**Comment by pixelzoom** [comment]
@jonathanolson On May 22 (immediately above), you indicated that you were done with #131 and were waiting to commit.  What is the status?

**Comment by jonathanolson** [comment]
It'll be a manual merging process to take my changes (made from the SVN version) and put them together (or replace) the git version that I was unaware of. Files in the gists linked.

**Comment by samreid** [comment]
Can the SVN version + your changes supercede the git version?  Is there anything SVN + your changes cannot do?

**Comment by jonathanolson** [comment]
Not moving through strings from Java, closing.

**Comment by pixelzoom** [comment]
> Not moving through strings from Java, closing.

One of us misunderstood, so I called @ariel-phet to confirm. We are still moving strings over from Java. We are then deleting any strings that contain HTML markup, because of the differing HTML support between Java and Scenery.

Reopening.

**Comment by pixelzoom** [comment]
Btw... I'm planning to use these tools for models-of-the-hydrogen-atom, so I'll reassign this to myself to bring them up to snuff.

@aaronsamuel137 How were HTML strings stripped out? Is there some code I can steal and add to these tools?

**Comment by aaronsamuel137** [comment]
No, I just used Intellij regex string replace with a few patterns, but can't remember the exact patterns off the top of my head.

**Comment by samreid** [comment]
Assigning to myself based on @pixelzoom's comments in https://github.com/phetsims/bending-light/issues/50

> Yes, we are still migrating strings from Java/Flash to HTML5. But I have not done so since strings were moved to babel. And the migration tools have not been updated to handle the migration changes. The plan was that whoever needed to do a migration next would address phetsims/chipper#132. Looks like @samreid is the lucky winner.

...

> Note that phetsims/chipper#132 was assigned to me because I started working on models-of-the-hydrogen-atom, and strings will need to be migrated for that. But since bending-light has just been published, it seems like migrating the strings has some urgency to it - and probably should have been done before publishing. At this point, you'll also need to be careful not to overwrite any translations that are done before you migrate the strings.

**Comment by samreid** [comment]
In the above commit, I've added a grunt task that ports the strings from Java .properties to babel .json (and nested in `value:{}`).  Outstanding work:
1. 2-term locales aren't supported at the moment (just a minor parse issue)
2. Checking in the 3rd party .properties parser to sherpa
3. Code cleanup/documentation (there is already some doc and eslint is passing, but I'd like to touch this up before review)
4. I was looking at bringing in other strings from other sims (say if "Slow Motion" was not translated into Danish in Bending Light (Java), but was translated in Energy Skate Park (Java), we could potentially use that string.  Could be false positives though, so we should be careful here or provide a way to opt-out.

**Comment by samreid** [comment]
Oh it looks like @jonathanolson may have already done something like this in Java in the gists mentioned in #131.

**Comment by samreid** [comment]
The above comment suggests to delete any translations that have html markup.  What pattern can I match for this?  <html>, <br>, anything else?  How can we prevent false positives?

Also, should we convert <br> to \n instead of deleting those translations?

**Comment by samreid** [comment]
Also, from the docs:

```js
  // This task has two modes:
  // (a) given a pre-existing JSON English file, port other languages based on the keys & values
  // This is useful when many strings have been changed/removed in porting to HTML5
  // (b) porting all of the files, including the English file.  This is performed if no English JSON
  // file can be found.
```

**Comment by pixelzoom** [comment]
We explicitly decided not to try to convert HTML to some other markup.  I would delete any string that contains equal numbers of '<' and '>', since it is likely to be HTML markup.  If we delete a few strings accidentally that are not HTML, that's preferable to accidentally keeping strings that contain markup.

**Comment by samreid** [comment]
I considered the following strategies for porting strings from Java .properties to babel .json.

(a) Port each file directly.  This would be a very simple solution that simply iterates over each Java .properties file, transforms directly to json and moves to babel

(b) Use the English translation file as "ground truth", and only port translations one key at a time, using a number of fallbacks to try to get as much coverage as possible.

Since I was working on Bending Light, I went with (b), though (a) would be more appropriate for a brand-new port.  Explaining some of the details regarding fallbacks in (b): 

Given an English string key/value pair "key1:jsonEnglishValue1":
1. If the translation file has a key "key1", then use its value for the translation
2. Otherwise, if the translation file has an exact string match for jsonEnglishValue1, then look up the corresponding key `key2` and look up the translated value for that (if any)
3. Otherwise if *any* english translation (from phetcommon or other sims) file has an exact string match for jsonEnglishValue1, then look up the corresponding key `key3' and look up the translated value for that (if any).  This is where homographs could provide incorrect results.  In the case of multiple different translation values for a single English string, my algorithm chooses the most frequently occurring value.

These fallback rules took the bending light "auto-port" coverage from about 50% to about 90%, since some of the keys had been renamed during porting.

It really did not take much time to create the initial version of this more sophisticated algorithm (b) but I'm wondering if it is too complex to bring to production and maintain.  On the other hand, given (b), then the implementation of (a) is simple--it just does a direct port of the english file, then runs the sophisticated algorithm (b) on the result.

It would be good to get feedback on whether we want to continue with (b) or just use (a).  Another option would be to port all of the strings directly, then write JSON-only tasks for dealing with strings (such as a task that throws away extraneous non-English translation key-value pairs.

I'm also curious about whether we should continue to do this on-demand, or whether we should do one last big batch of all properties and xml.  Pro: this would simplify our ongoing task.  Con: this would flood babel (or some other repo) and we may miss out on contributions made to our Java/Flash sims between now and when we port the sim to HTML5.

Unassigning until I get feedback from @jonathanolson @jbphet @pixelzoom or @jessegreenberg.

**Comment by pixelzoom** [comment]
> It would be good to get feedback on whether we want to continue with (b) or just use (a).

This is ultimately a question for @ariel-phet - are you willing to maintain a more complicated piece of code in exchange for a higher percentage of ported strings?  

My feeling is that we should be doing something simple, and relying on the translator to fill in the blanks.  And option (a) has served us well to date. That said...

> This is where homographs could provide incorrect results. In the case of multiple different translation values for a single English string, my algorithm chooses the most frequently occurring value.

Unless there's some reason to expect this algorithm to be 100% accurate, my vote would be to toss homographs.

**Comment by samreid** [comment]
Jan 7 2015: Meeting consensus: it would be good to move these as a batch so we can abandon this tool.  Newly submitted strings to old Java sims would be omitted.

@oliver-phet says translations coming in to Java are becoming more rare.

**Comment by samreid** [comment]
Do not try to grab strings from other contexts, there is too much probability of a mismatch.

**Comment by pixelzoom** [comment]
Will need to do something similar for Flash strings.

**Comment by samreid** [comment]
A reminder for whoever works on this, to strip out the HTML.

@pixelzoom says just check if there is a "<" character, then presume it is HTML and toss it out.

@ariel-phet says @samreid don't spend too much time on this.  If it would be too time consuming, reassign to @ariel-phet

**Comment by samreid** [comment]
I don't think I should proceed until discussing @jonathanolson's gists in #131, reassigning to @jonathanolson until then.

**Comment by samreid** [comment]
@jonathanolson comments are in #131 and I am good to proceed.

**Comment by samreid** [comment]
Most of the prior work I'm building on is in #131 so I'm working in that issue.  I've already committed the ported Java strings and will work on XML next.  I think this issue can be closed and we'll continue with XML in #131.
